### PR TITLE
server: add hot region table/index api

### LINKED ISSUE #7870: [FR] Use SQL or HTTP-API to get the hotspot
## Feature Request

**Is your feature request related to a problem? Please describe:**

It is not easy to find out which table/index is the hotspot when the database is slow.

**Describe the feature you'd like:**

We could get the hotspot through `pd-ctl`. But it is inconvenient for the users to got insight which table/index is the hotspot. They have to decode the region start-key/end-key to get table/index id, then use tidb's HTTP API to locate the table/index name.

We'd better provide a one-click way to show which table/index is the hotspot.

**Describe alternatives you've considered:**

No

**Teachability, Documentation, Adoption, Migration Strategy:**

When the TiDB cluster is slow, user could easily find out the root cause.

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

fixes #7870 

### What is changed and how it works?

add http api to query all hot table/index

### Check List

Tests 

 - Manual test (add detailed scripts or steps below)

1. setup PD/TiKV/TiDB/Promethus/Grafana in local machine
2. use sysbench to prepare data, this will got hot region
3. then acess http://0.0.0.0:10080/regions/hot to see hot table/idx list in desc order

```
{
  "read": [
    
  ],
  "write": [
    {
      "db_name": "sbtest1",
      "table_name": "sbtest13",
      "index_name": "",
      "flow_bytes": 220718,
      "max_hot_degree": 12,
      "region_count": 1
    },
    {
      "db_name": "sbtest1",
      "table_name": "sbtest1",
      "index_name": "",
      "flow_bytes": 134524,
      "max_hot_degree": 11,
      "region_count": 1
    },
    {
      "db_name": "sbtest1",
      "table_name": "sbtest12",
      "index_name": "",
      "flow_bytes": 115753,
      "max_hot_degree": 11,
      "region_count": 1
    },
    {
      "db_name": "sbtest1",
      "table_name": "sbtest4",
      "index_name": "",
      "flow_bytes": 214674,
      "max_hot_degree": 10,
      "region_count": 1
    }
  ]
}
```
then we can using `jq` tool to filter the result

Code changes

 - Has exported http_api and metric change
 - Has exported variable/fields change

Side effects

 - No

Related changes

 - N/A

### Remain question

- code refine

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7890)

### CODE REVIEW

**Comment by ngaut** [comment]
Coool

**Comment by lysu** [comment]
@shenli PTAL, I just update the description

File: server/http_status.go [hunk 1]
```diff
@@ -67,11 +67,17 @@ func (s *Server) startHTTPServer() {
 		router.Handle("/tables/{db}/{table}/stop-scatter", tableHandler{tikvHandlerTool, opStopTableScatter})
 		router.Handle("/tables/{db}/{table}/disk-usage", tableHandler{tikvHandlerTool, opTableDiskUsage})
 		router.Handle("/regions/meta", regionHandler{tikvHandlerTool}).Name("RegionsMeta")
+		router.Handle("/regions/hot", regionHandler{tikvHandlerTool}).Name("RegionHot")
```
Comment (jackysp) [inline server/http_status.go]: Don't forget to update the document of http api.

**Comment by lysu** [comment]
@jackysp @tiancaiamao  PTAL if free, I have remove promethus exporter code to avoid too many label values problem, now just a simple http api.

File: docs/tidb_http_api.md [hunk 2]
```diff
@@ -20,6 +20,12 @@
     curl http://{TiDBIP}:10080/regions/meta
     ```
 
+1. Get the table/index of hot regions
+
+    ```shell
+    curl http://0.0.0.0:10080/regions/hot
```
Comment (jackysp) [inline docs/tidb_http_api.md]: It's better to use {TiDBIP}.

File: server/http_handler.go [hunk 3]
```diff
@@ -304,6 +308,151 @@ func (t *tikvHandlerTool) getAllHistoryDDL() ([]*model.Job, error) {
...
+func (t *tikvHandlerTool) fetchHotRegion(rw string) (map[uint64]regionMetric, error) {
+	etcd, ok := t.store.(domain.EtcdBackend)
+	if !ok {
+		return nil, errors.New("not implemented")
+	}
+	pdHosts := etcd.EtcdAddrs()
+	if len(pdHosts) == 0 {
+		return nil, errors.New("pd unavailable")
+	}
+	req, err := http.NewRequest("GET", protocol+pdHosts[0]+rw, nil)
```
Comment (tiancaiamao) [inline server/http_handler.go]: Why not use `http.Get` ?

File: server/http_handler.go [hunk 4]
```diff
@@ -304,6 +308,151 @@ func (t *tikvHandlerTool) getAllHistoryDDL() ([]*model.Job, error) {
...
+		return nil, errors.Trace(err)
+	}
+	defer func() {
+		err = resp.Body.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+	var regionResp storeHotRegionInfos
+	err = json.NewDecoder(resp.Body).Decode(&regionResp)
```
Comment (tiancaiamao) [inline server/http_handler.go]: Will the response struct change?
How about decoding as `map[string]interface{}`

File: server/http_handler.go [hunk 5]
```diff
@@ -304,6 +308,151 @@ func (t *tikvHandlerTool) getAllHistoryDDL() ([]*model.Job, error) {
...
+		for _, region := range hotRegions.RegionsStat {
+			metric[region.RegionID] = regionMetric{FlowBytes: region.FlowBytes, MaxHotDegree: region.HotDegree}
+		}
+	}
+	// TODO: not leader peer....
+	return metric, nil
+}
+
+func (t *tikvHandlerTool) fetchRegionTableIndex(metrics map[uint64]regionMetric) (map[tblIndex]regionMetric, error) {
+	s, err := session.CreateSession(t.store.(kv.Storage))
```
Comment (tiancaiamao) [inline server/http_handler.go]: The session is created just to get the infoschema?
There should be better ways to achieve the goal.

File: server/http_handler.go (hunk 3 above)
Comment (lysu) [reply server/http_handler.go]: never use `http.Get`, we should control timeout to call other resources.

File: server/http_handler.go (hunk 4 above)
Comment (lysu) [reply server/http_handler.go]: API change need backward compatibility, this method can handle compatibility change.

and use `interface{}` doesn't solve question, later type assert or get field will failure even you use weak type.

**Comment by lysu** [comment]
/run-all-tests

@tiancaiamao @jackysp  PTAL if free, thx

**Comment by lysu** [comment]
@tiancaiamao  PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**